### DISCUSSION TITLE
How to run sequential async calls with class methods

### DISCUSSION THREAD

**Discussion by zilch42** [zauberzeug/nicegui#2358] [Q&A] [answered]
### Question

Hi there, 

I'm struggling with some async stuff, but also having issues when I try to do the same tasks without async at all, if those tasks take too long. 

Here's a small app that downloads some csv's from the `nycflights13` dataset, loads them into a DataModel class, and does a calculation with the data using a class method. In my thinking, the download step is `io_bound` and the loading and calculation steps are `cpu_bound`. However, if I try and run those 3 steps sequentially with async calls, it looks like the last a calculation step tries to run before the data has finished loading. 

```python
import math
from io import StringIO

from nicegui import run, ui
import pandas as pd
import requests

# raw data from R Tidyverse package on github - NYC flights 13
URLs = {'planes': 'https://raw.githubusercontent.com/tidyverse/nycflights13/main/data-raw/planes.csv',
        'airports': 'https://raw.githubusercontent.com/tidyverse/nycflights13/main/data-raw/airports.csv'}


class DataModel:
    """Store data and perform calculations"""

    def __init__(self):
        self.planes = None
        self.airports = None
        self.mean = None

    def load(self, planes, airports):
        """load csv responses into pandas dfs"""
        self.mean = None

        for i in range(int(repeats.value) + 1):
            self.planes = pd.read_csv(StringIO(planes.content.decode('utf-8')))
            self.airports = pd.read_csv(StringIO(airports.content.decode('utf-8')))
            # other transformations, etc

    def heavy_computation(self):
        """Run some heavy computation and save a value in the model"""
        for i in range(int(repeats.value) + 1):
            self.mean = self.planes.year.apply(lambda x: math.sqrt(x+1)).mean()

# other functions
def download_data(URLs):
    """get csv data from github"""
    for i in range(int(repeats.value) + 1):
        response = {k: requests.get(v, timeout=3) for k, v in URLs.items()}
    return response


async def handle_click_normal():
    """Download data, load into model, perform a calculation, no Async"""
    response = download_data(URLs)
    myDataModelNormal.load(**response)
    myDataModelNormal.heavy_computation()
    ui.notify("Done without Async")


async def handle_click_async():
    """Download data, load into a DataModel, perform a calculation, all async"""
    response = await run.io_bound(download_data, URLs)
    await run.cpu_bound(myDataModelAsync.load, **response)
    await run.cpu_bound(myDataModelAsync.heavy_computation)
    ui.notify("Done Asynchronously")


def reset():
    myDataModelNormal.__init__()
    myDataModelAsync.__init__()


# setup data models
myDataModelNormal = DataModel()
myDataModelAsync = DataModel()


# page UI
with ui.row():
    ui.button('Process NYCflights13 Normally', on_click=handle_click_normal)
    ui.label().bind_text_from(myDataModelNormal, 'mean',
                              backward=lambda x: "no data" if not x else str(x) + " average year")
with ui.row():
    ui.button('Process NYCflights13 Async', on_click=handle_click_async)
    ui.label().bind_text_from(myDataModelAsync, 'mean',
                              backward=lambda x: "no data" if not x else str(x) + " average year")

ui.button("Reset models", on_click=reset)

# times to repeat so that calculations take longer
repeats = ui.number("Times to repeat (so calculations take longer): ",
                    value=0, min=0).style("width: 400px")
ui.label().bind_text_from(repeats, 'value')

ui.run()
```

Clicking the `PROCESS NYCFLIGHTS13 NORMALLY` button produces a result, 

![image](https://github.com/zauberzeug/nicegui/assets/39004650/8dd150c7-79ba-44a7-a947-384106082499)

while clicking `PROCESS NYCFLIGHTS13 ASYNC` produces the following error.

```
'NoneType' object has no attribute 'year'
concurrent.futures.process._RemoteTraceback:
"""
Traceback (most recent call last):
  File "C:\Users\path\lib\concurrent\futures\process.py", line 246, in _process_worker   
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "c:\Users\abb064\OneDrive - CSIRO\R\AiCE\csiro-horizon-scanning\apps\nicegui\async\async.py", line 33, in heavy_computation
    self.mean = self.planes.year.apply(lambda x: math.sqrt(x+1)).mean()
AttributeError: 'NoneType' object has no attribute 'year'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\path\lib\site-packages\nicegui\events.py", line 406, in wait_for_result 
    await result
  File "c:\Users\project\apps\nicegui\async\async.py", line 56, in handle_click_async
    await run.cpu_bound(myDataModelAsync.heavy_computation)
  File "C:\Users\path\lib\site-packages\nicegui\run.py", line 28, in cpu_bound
    return await _run(process_pool, callback, *args, **kwargs)
  File "C:\Users\path\lib\site-packages\nicegui\run.py", line 18, in _run
    return await loop.run_in_executor(executor, partial(callback, *args, **kwargs))
AttributeError: 'NoneType' object has no attribute 'year'
```

How can I make this run asynchronously correctly?

Another curiosity is that if I increase the number of repeats (just to make the calculations take longer), at a certain point, running normally without with any async calls starts to break down. I don't get any error printed, but I don't get any result either. 

The breaking point might depend on your system but for me, 0 repeats is obviously fine, 2 is fine, 5 is iffy, at 10 I never get a result. 10 repeats should take < 20sec. 

Any idea what's going on there?

**Comment by rodja** [comment, marked as answer]
If you run blocking tasks on the main loop (eg. not async) you will halt everything. So your observation with "Process NYCflights13 Normally" is expected. See https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates.

And the `run.cpu_bound` tasks should not access class properties. As explained in https://github.com/zauberzeug/nicegui/discussions/2221#discussioncomment-7920864, `run.cpu_bound` needs to execute the function in a separate process. For this it needs to transfer the whole state of the passed function to the process (which is done with pickle). It is encouraged to create static methods (or free functions) which get all the data as simple parameters (eg. no class/ui logic) and return the result (instead of writing it in class properties or global variables).

Because it is such a valuable information we will add it to the doc string in the next release: https://github.com/zauberzeug/nicegui/pull/2363

**Reply by ariahrism** [reply]
Hi! I'm really struggling with this too actually. I'm using async functions, and `await` wherever I can, but I can't seem to understand the best practice.

My use case: 
I have a search box to fetch data. After fetching the data I want to handle it. No issues, nothing hangs.
But then when I go to iterate over the data, parts of the UI stall the other parts of the UI. 
It's fine that some parts of the UI load slow. But I'd like it to not freeze up the entire page.
Simple example below. Perhaps we can use it in documentation as an example... if you think its a good idea?


```python
from nicegui import ui, run
import time
import asyncio

row1 = ui.row()
row2 = ui.row()
row3 = ui.row()

def searched(): #  Have tried using async and await in various combinations 
    populate_row1()
    populate_row2()
    populate_row3()
        #  Have tried run.cpu_bound and io_bound
        #  Have tried asyncio.create_task

def populate_row1():
    with row1:
        ui.label('Search details 1')

def populate_row2():
    with row2:
        ui.label('Search details 2')

def populate_row3(search_results=[1,2]): #  Continues to stall rendering of row1 and row2
    with row3:
        for things in search_results:
            with ui.card():
                ui.label(f'Lots of UI elements to show {things}')
            time.sleep(15)  #  Simulate processing of many ui elements like ui.cards()

ui.button('Search', on_click=searched)

ui.run()
```

What would be the best practice? I know it's been explained, but I'm really not able to comprehend how to get it out of the main thread.

**Reply by rodja** [reply]
`time.sleep(15)` is blocking the main loop. And similar creating hundreds or more UI elements will block the main thread. The best way would be to not create all elements directly but only when they come into view. Like we do in the [infinite scroll example](https://github.com/zauberzeug/nicegui/blob/main/examples/infinite_scroll/main.py). If that is no option, you could place `await time.sleep(0)` calls inside your loop as I described in https://github.com/zauberzeug/nicegui/issues/2290#issuecomment-1878707295.

**Reply by ariahrism** [reply]
Thank you for that!
> time.sleep(15) is blocking the main loop. 

Yeah I tried various ways to get it to work including `await run.cpu_bound`. The example above was just showcasing the issue, providing comments for what I tried.

I still don't understand it fully, but I'm closer. After sleeping on it, I thought to also try using `threading` to make 3 different threads listening to the search event. I couldn't really get it... 

In the end, your solution works as follows:

```python
from nicegui import ui, run
import time
import asyncio

row1 = ui.row()
row2 = ui.row()
row3 = ui.row()

search_results = None
async def searched():
    global search_results
    search_results = [1,2]
    # await populate_row1()
    # await populate_row2()
    # await populate_row3()
    await asyncio.gather( # Run all 3 at the same time instead of sequentially
            populate_row1(),
            populate_row2(),
            populate_row3(),
        )

async def populate_row1():
    with row1:
        ui.label('Search details 1')

async def populate_row2():
    with row2:
        ui.label('Search details 2')

async def populate_row3():
    with row3:
        for things in range(100):# things in search_results:
            await asyncio.sleep(2)
            with ui.card():
                ui.label(f'Lots of UI elements to show {things}')
            await run.cpu_bound(time.sleep, 2)
            with ui.card():
                ui.label(f'Lots of CPU bound work to do {things}')

ui.button('Search', on_click=searched)

ui.run(show=False)
```

I also tried this to listen for the search, but something tells me it's less stable. Although I can't say that because I don't exactly understand the differences between the two implementations.
Still it works and it's interesting:

```python
from nicegui import ui, run, app
import time
import asyncio

search_event = asyncio.Event()
search_event.clear()

row1 = ui.row()
row2 = ui.row()
row3 = ui.row()

search_results = None
def searched():
    global search_results
    search_results = [1,2]
    search_event.set()  # Signal that new data is available

async def populate_row1():
    while True:
        await search_event.wait()  # Wait for new data
        search_event.clear()  # Reset the event, dangerous because other tasks may be waiting on this event
        with row1:
            ui.label('Search details 1')

async def populate_row2():
    while True:
        await search_event.wait()  # Wait for new data
        search_event.clear()  # Reset the event, dangerous because other tasks may be waiting on this event
        with row2:
            ui.label('Search details 2')

async def populate_row3():
    while True:
        await search_event.wait()  # Wait for new data
        search_event.clear()  # Reset the event, dangerous because other tasks may be waiting on this event
        with row3:
            for things in range(100):# things in search_results:
                await asyncio.sleep(5)
                with ui.card():
                    ui.label(f'Lots of UI elements to show {things}')
                await run.cpu_bound(time.sleep, 2)
                with ui.card():
                    ui.label(f'Lots of CPU bound work to do {things}')
                
def start_tasks():
    asyncio.create_task(populate_row1())
    asyncio.create_task(populate_row2())
    asyncio.create_task(populate_row3())

app.on_startup(start_tasks)

ui.button('Search', on_click=searched)

ui.run(show=False)
```

GPT tells me this: 😆

>Recommendation:
> - The first implementation is generally more readable and aligns with common async/await patterns in Python.
> - If you have more complex synchronization requirements or multiple tasks waiting for the same event, the second implementation might be more suitable.
> 
> In terms of stability and performance, both should work well for your use case.

**Comment by zilch42** [comment]
Thanks @rodja, that's helped clear things up for me. Just for confirmation, and for anyone else finding this and wanting a working example, would this be the best way to do things if I insist on using a class? 

Main changes: 
- do any processing in `cpu_bound` static methods and return the value
- save the data to class in the main thread (not bound)

```python
import math
from io import StringIO

from nicegui import run, ui
import pandas as pd
import requests

# raw data from R Tidyverse package on github - NYC flights 13
URLs = {'planes': 'https://raw.githubusercontent.com/tidyverse/nycflights13/main/data-raw/planes.csv',
        'airports': 'https://raw.githubusercontent.com/tidyverse/nycflights13/main/data-raw/airports.csv'}


class DataModel:
    """Store data and perform calculations"""

    def __init__(self):
        self.planes = None
        self.airports = None
        self.mean = None

    def load_data(self, planes, airports):
        """load prepared data"""
        self.mean = None
        self.planes = planes
        self.airports = airports

    @staticmethod
    def prepare_data(dic):
        """convert csv responses into pandas dfs"""
        for i in range(int(repeats.value) + 1):
            data = {k: pd.read_csv(StringIO(v.content.decode('utf-8'))) for k, v in dic.items()}
        return data

    @staticmethod
    def heavy_computation(df):
        """Run some heavy computation and save a value in the model"""
        for i in range(int(repeats.value) + 1):
            mean = df.year.apply(lambda x: math.sqrt(x+1)).mean()
        return mean


def download_data(URLs):
    """get csv data from github"""
    for i in range(int(repeats.value) + 1):
        response = {k: requests.get(v, timeout=3) for k, v in URLs.items()}
    return response


async def handle_click_async():
    """Download data, load into a DataModel, perform a calculation, all async"""
    response = await run.io_bound(download_data, URLs)
    prepared_data = await run.cpu_bound(myDataModelAsync.prepare_data, response)
    
    # save data to class in main thread - minimal processing
    myDataModelAsync.load_data(**prepared_data)
    
    myDataModelAsync.mean = await run.cpu_bound(myDataModelAsync.heavy_computation, myDataModelAsync.planes)
    
    ui.notify("Done Asynchronously")


def reset():
    myDataModelAsync.__init__()


# setup data models
myDataModelAsync = DataModel()


# page UI
with ui.row():
    ui.button('Process NYCflights13 Async', on_click=handle_click_async)
    ui.label().bind_text_from(myDataModelAsync, 'mean',
                              backward=lambda x: "no data" if not x else str(x) + " average year")

ui.button("Reset models", on_click=reset)

# times to repeat so that calculations take longer
repeats = ui.number("Times to repeat (so calculations take longer): ",
                    value=0, min=0).style("width: 400px")
ui.label().bind_text_from(repeats, 'value')

ui.run()
```

**Comment by BlankAdventure** [comment]
I've been running into the exact same issue. I'm trying to write a UI to wrap the functionality of an existing simulation library, which relies on heavy OOP. Unfortunately I don't have the freedom to rewrite it all into a function-first/static method structure. 

Are there any other workarounds? I was also thinking about the possibility of using an async queue or something (basically push the heavy calculation onto a queue with an async worker loop running to retrieve and process it)

**Reply by rodja** [reply]
@BlankAdventure are you asking about a solution to make a cpu intensive class usable in an async program? This is rather broad and may attract a broader community if asked on StackOverflow or Reddit...

**Reply by BlankAdventure** [reply]
Yes, but you're right that its a broad topic - I've seen some suggestions on StackOverflow to use dill instead of pickle with multiprocessing: https://stackoverflow.com/questions/19984152/what-can-multiprocessing-and-dill-do-together 

As for something that just works there is some interesting discussion here:
https://stackoverflow.com/questions/43241221/how-can-i-wrap-a-synchronous-function-in-an-async-coroutine

I've tried the async wrap decorator with some nicegui apps and and it seems to work - though not sure if it would be considered a robust solution.

**Reply by rodja** [reply]
The `@wrap` decorator has the same issue as our `run.cpu_bound`: the class with all its members needs to be serialized and brought to another process. All changes you make there are tricky to sync back to the main process. An alternative is to create all your synchronous classes only in a separate sub-process and define a simpler communication protocol which can be used from the async context. We have done this for our PathPlanner in [RoSys](https://rosys.io) (our robotic framework on top of NiceGUI):

https://github.com/zauberzeug/rosys/blob/e38bcfe344849fb84ae1eb9a2f90b3e3c1ecab07/rosys/pathplanning/path_planner.py#L31

**Reply by BlankAdventure** [reply]
Yes, you are correct. I realized I missed an important piece -- in my case, I was receiving the pickle error not because I was using cpu_bound with class methods (which is okay, so long as they're defined at the top level), but instead because said code had a bunch of lambdas hiding inside. So that's what broke it. The @wrap decorator resolved this problem.
